Reorganizing "Communicating Network Status"

gcolten opened the pull request:
Incomplete, but illustrates the direction I'm heading

gcolten commented:
This is a pretty opinionated review. I wrote it to sound like me, so I can appreciate if you all would like to change some of the style.

I welcome all feedback, as well as any concerns with the technical descriptions themselves. I definitely recommend double-checking the technical accuracy - I would _not_ consider myself an expert here. I just read the GCC draft and thought this could benefit from a summary of what I understood.

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -197,35 +197,63 @@ Heuristics that model the network behavior and tries to predict it is known as B
+2. Communicate network characteristics between sender and receiver
+
+RTP/RTCP has three different approaches to address this problem. They all have their pros and cons,
+and generally each generation has improved over its precessors. Which implementation you use will
```
Google seems to say that `precessors` is an obsolete spelling, `predecessors`?

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -197,35 +197,63 @@ Heuristics that model the network behavior and tries to predict it is known as B
+RTCP messages are defined in [RFC 3550](https://tools.ietf.org/html/rfc3550#section-6.4), and are
+responsible for communicating network status between endpoints. Receiver Reports focuses on
+communicating qualities about the network (including packet loss, round-trip time, and jitter), and
+it pairs with other algorithms that are then responsible for estimating avilable bandwidth based
+off of these reports.
```
Nit: "off of" is correct, but sounds a bit funny. Maybe "base on these reports."?

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -197,35 +197,63 @@ Heuristics that model the network behavior and tries to predict it is known as B
+
+Sender and Receiver reports (SR and RR) together paint a picture of the network quatlity. They are
+sent on a schedule for each SSRC, and they are the inputs used when estimating available
+bandwidth. Those estimates take place on at the sender after receiving the RR data, containing the
```
Nit: How about "The estimates are made by the sender after receiving the RR data,"?

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+* If packet loss is below 2%, the bandwidth estimate is increased
+
+Packet loss measurements are taking frequently, and rely on metadata inside each packet to be
+able to infer when packets are missing. These percentages are evaluated over time windows of around
+0.5-1.0 seconds.
```
Nit: Maybe simplify this even more? "These percentages are evaluated over time windows of about a second."?

mogren reviewed (commented):
Thanks, this is great! I'll read through some more later, need to take a break. Just leaving some comments for now.

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+packet arrival time. This delay-based controller aims to identify when network links are becoming
+increasingly congested, and may reduce bandwidth estimates even before packet loss occurs. The
+theory is that the busiest network interface along the path will continue queuing up packets up
+until the interface runs out of capacity inside its buffers. If that interface continues receiving
```
Nit: "continues receiving" -> "continues to receive"

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+increasingly congested, and may reduce bandwidth estimates even before packet loss occurs. The
+theory is that the busiest network interface along the path will continue queuing up packets up
+until the interface runs out of capacity inside its buffers. If that interface continues receiving
+more traffic than it is able to send, it will be forced to drop all of the next packets it receives
+until more buffer spaces become available. This type of packet loss can be disruptive to all
```
Nit: "the following packets it receives until more buffer space become available."

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+frames is smaller than the arrival time difference between the second and third frame), that is
+considered evidence of increased queue depth on the connecting network interfaces and presumed to
+be caused by network congestion. (Note: GCC is smart enough to control these estimates for
+fluctuations in frame byte sizes.) GCC refines its latency measurements using [Kalman
```
Nit: "using a Kalman..."

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+considered evidence of increased queue depth on the connecting network interfaces and presumed to
+be caused by network congestion. (Note: GCC is smart enough to control these estimates for
+fluctuations in frame byte sizes.) GCC refines its latency measurements using [Kalman
+filter](https://en.wikipedia.org/wiki/Kalman_filter) and takes very many measurements of network
```
Nit: "very many measurements" -> "a lot of measurements"

mogren commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+estimates to test out higher load values.
+
+#### TMMBR, TMMBN, and REMB
+After identifyign an estimate for available inbound bandwidth, receivers then communicate these
```
"identifyign" ->  "identifying"

gcolten commented:
Great catches!  I evidently let a few spelling mistakes through, but those should all be fixed now.

I believe I addressed all of the feedback shared so far.

mogren reviewed (approved):
Great stuff, thanks a lot! LGTM! 🎉

mogren commented:
@mengelbart @Sean-Der Any additional comments? 😄

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+To do so, GCC tries to infer increases in queue depth by measuring subtle increases in in round
+trip time. It records what's called the frame inter-arrival time, `t(i) - t(i-1)`, as the
+difference in arrival time of two groups of packets (generally, consecutive video frames). If the
+inter-arrival time increases over time, (as in, the arrival time difference between the first two
```
`inter-arrival time relative to departure time`, as there might be differences in the deltas between the send times of packets. I think `06-twcc.png` would be a nice illustration here, too.

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+frames is smaller than the arrival time difference between the second and third frame), that is
+considered evidence of increased queue depth on the connecting network interfaces and presumed to
+be caused by network congestion. (Note: GCC is smart enough to control these estimates for
+fluctuations in frame byte sizes.) GCC refines its latency measurements using a [Kalman
```
It may be helpful to clarify why the Kalman (or any other) filter is used. The measured delay variations include some noise (e.g. jitter) and the filter is used to produce a more accurate estimation of the real value from the noisy measurements. Maybe it is also worth mentioning, that one could other methods instead of the Kalman filter (I think e.g. Signal and libwebrtc use linear regression).

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -274,40 +351,62 @@ For example, using the x264 encoder with `tune=zerolatency` can significantly de
 
 ### Transport Wide Congestion Control
-Transport Wide Congestion Control is the latest development in RTCP network status communication.
+Transport Wide Congestion Control is the latest development in RTCP network status
```
Since last year there is also RFC 8888, which defines a similar feedback message for the same purpose. I think TWCC is still widely used due to being implemented in Browsers, but it may be worth mentioning, that RFC 8888 also exists.

gcolten replied:
That could be valuable, but I'm not sure I should be the one to add that.

Having not seen that RFC referenced before, I'm not sure what context to give it.

mogren replied:
I had not seen it either, thanks for the tip @mengelbart. I filed #246 to track it. Might be easier to do in a follow up PR.

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -274,40 +351,62 @@ For example, using the x264 encoder with `tune=zerolatency` can significantly de
-If the receive delays increase, it means network congestion is happening, and the sender must act on it.
+With TWCC, the receiver lets the sender know the arrival time of each packet. This is enough
+information for the sender to measure inter-packet arrival delay variation, as well as identifying
+exactly which packets were dropped. With this data being exchanged frequently, the sender able to
```
`exactly which packets were dropped` sounds too extreme, it can actually only identify which packets have not (yet) arrived. Some packets may arrive late/re-ordered e.g. due to jitter.

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -274,40 +351,62 @@ For example, using the x264 encoder with `tune=zerolatency` can significantly de
-
-![TWCC with delay](../images/06-twcc.png "TWCC with delay")
+The sender keeps track of sent packets, their sequence numbers, sizes and timestamps.  When the
+sender receives RTCP messages from the receiver, it compares the send inter-packet delays with
```
Instead of explaining inter-arrival delays again, I would suggest just linking back to GCC.

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -274,40 +351,62 @@ For example, using the x264 encoder with `tune=zerolatency` can significantly de
-The most deployed implementation is "A Google Congestion Control Algorithm for Real-Time Communication" defined in [draft-alvestrand-rmcat-congestion](https://tools.ietf.org/html/draft-alvestrand-rmcat-congestion-02).
-It can run in two passes. First a "loss based" pass that just uses Receiver Reports. If TWCC is available, it will also take that additional data into consideration.
-It predicts the current and future network bandwidth by using a [Kalman filter](https://en.wikipedia.org/wiki/Kalman_filter).
+![TWCC with delay](../images/06-twcc.png "TWCC with delay")
```
This diagram could be helpful in the GCC section, too, see comment above.

mengelbart commented on content/docs/06-media-communication.md:
```diff
@@ -274,40 +351,62 @@ For example, using the x264 encoder with `tune=zerolatency` can significantly de
+mobile devices, this means those clients can benefit from algorithm enhancements without having to
+await standardization or browser updates (which can take quite a long time to be widely available).
+
+<!-- Missing: protocol for estimating bandwidth that accompanies TWCC -->
```
GCC works with TWCC, NADA and SCReAM should in theory work, too, but I don't know if that is implemented anywhere.

mengelbart reviewed (commented):
Sorry I missed this earlier, @mogren @gcolten ! I made some comments/suggestions now. Thanks for writing this down, I think it will be very helpful in learning/understanding bandwidth estimation! :)

gcolten commented on content/docs/06-media-communication.md:
```diff
@@ -234,32 +262,81 @@ Round-trip time in plain English:
+theory is that the busiest network interface along the path will continue queuing up packets up
+until the interface runs out of capacity inside its buffers. If that interface continues to receive
+more traffic than it is able to send, it will be forced to drop all of the next packets it receives
+until more buffer space becomes available. This type of packet loss can be disruptive to all
```
I guess it depends on your definition of "disruptive".

Sure, TCP is capable of recovering, but even TCP doesn't prefer to have packets dropped. High latency connections or those without Selective Acknowledgement enabled may stall awaiting the retransmit timeout timer.

So I would certainly say that packet loss is disruptive to TCP throughput, even if it isn't necessarily disruptive to all applications running TCP.

gcolten commented:
Alrighty. Thanks for the thoughts, @mengelbart! I do think the reorganization you suggested is valuable; I just hope that I didn't introduce any other issues as a result... :)
